File: PagingRemoteMediatorJava/app/src/main/java/com/websarva/wings/android/pagingremotemediatorjava/MainActivity.java
package com.websarva.wings.android.pagingremotemediatorjava;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.paging.PagingData;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.websarva.wings.android.pagingremotemediatorjava.data.local.Phone;
import com.websarva.wings.android.pagingremotemediatorjava.databinding.ActivityMainBinding;
import com.websarva.wings.android.pagingremotemediatorjava.ui.MainViewModel;

public class MainActivity extends AppCompatActivity {
	private ActivityMainBinding _activityMainBinding;
	private MainViewModel _mainViewModel;
	private PhoneListAdapter _phoneListAdapter;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		EdgeToEdge.enable(this);
		_activityMainBinding = ActivityMainBinding.inflate(getLayoutInflater());
		View contentView = _activityMainBinding.getRoot();
		setContentView(contentView);
		ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
			Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
			v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
			return insets;
		});
		ViewModelProvider provider = new ViewModelProvider(MainActivity.this);
		_mainViewModel = provider.get(MainViewModel.class);

		_activityMainBinding.rvPhoneList.setLayoutManager(new LinearLayoutManager(MainActivity.this));
		DividerItemDecoration decoration = new DividerItemDecoration(MainActivity.this, DividerItemDecoration.VERTICAL);
		_activityMainBinding.rvPhoneList.addItemDecoration(decoration);
		_phoneListAdapter = new PhoneListAdapter(new PhoneComparator());
		_activityMainBinding.rvPhoneList.setAdapter(_phoneListAdapter);

		LiveData<PagingData<Phone>> phoneListLiveData = _mainViewModel.getPhoneListLiveData();
		phoneListLiveData.observe(MainActivity.this, new PhoneListObserver());
	}

	private class PhoneListObserver implements Observer<PagingData<Phone>> {
		@Override
		public void onChanged(PagingData<Phone> phonePagingData) {
			_phoneListAdapter.submitData(getLifecycle(), phonePagingData);
		}
	}

	private class PhoneViewHolder extends RecyclerView.ViewHolder {
		private final TextView _tvId;
		private final TextView _tvPhoneNo;

		public PhoneViewHolder(View itemView) {
			super(itemView);
			_tvPhoneNo = itemView.findViewById(android.R.id.text1);
			_tvId = itemView.findViewById(android.R.id.text2);
		}

		public void bind(Phone item) {
			if(item != null) {
				_tvId.setText(String.valueOf(item.id));
				_tvPhoneNo.setText(item.phoneNo);
			}
		}
	}

	private class PhoneListAdapter extends PagingDataAdapter<Phone, PhoneViewHolder> {
		public PhoneListAdapter(@NonNull DiffUtil.ItemCallback<Phone> diffCallback) {
			super(diffCallback);
		}

		@NonNull
		@Override
		public PhoneViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			LayoutInflater inflater = LayoutInflater.from(MainActivity.this);
			View row = inflater.inflate(android.R.layout.simple_list_item_2, parent, false);
			return new PhoneViewHolder(row);
		}

		@Override
		public void onBindViewHolder(@NonNull PhoneViewHolder holder, int position) {
			Phone item = getItem(position);
			holder.bind(item);
		}
	}

	private class PhoneComparator extends DiffUtil.ItemCallback<Phone> {
		@Override
		public boolean areItemsTheSame(@NonNull Phone oldItem, @NonNull Phone newItem) {
			return oldItem.id == newItem.id;
		}

		@Override
		public boolean areContentsTheSame(@NonNull Phone oldItem, @NonNull Phone newItem) {
			return oldItem.id == newItem.id && oldItem.phoneNo.equals(newItem.phoneNo);
		}
	}
}

File: PagingRemoteMediatorKotlin/app/src/main/java/com/websarva/wings/android/pagingremotemediatorkotlin/MainActivity.kt
package com.websarva.wings.android.pagingremotemediatorkotlin

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.websarva.wings.android.pagingremotemediatorkotlin.data.local.Phone
import com.websarva.wings.android.pagingremotemediatorkotlin.databinding.ActivityMainBinding
import com.websarva.wings.android.pagingremotemediatorkotlin.ui.MainViewModel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
	private val _mainViewModel by viewModels<MainViewModel>()
	private lateinit var _activityMainBinding: ActivityMainBinding

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		enableEdgeToEdge()
		_activityMainBinding = ActivityMainBinding.inflate(layoutInflater)
		setContentView(_activityMainBinding.root)
		ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
			val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
			v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
			insets
		}

		_activityMainBinding.rvPhoneList.layoutManager = LinearLayoutManager(this@MainActivity)
		val decoration = DividerItemDecoration(this@MainActivity, DividerItemDecoration.VERTICAL)
		_activityMainBinding.rvPhoneList.addItemDecoration(decoration)
		val adapter = PhoneListAdapter(PhoneComparator())
		_activityMainBinding.rvPhoneList.adapter = adapter

		lifecycleScope.launch {
			repeatOnLifecycle(Lifecycle.State.STARTED) {
				_mainViewModel.phoneListFlow.collectLatest {
					adapter.submitData(it)
				}
			}
		}
	}

	private inner class PhoneViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		private val _tvId: TextView
		private val _tvPhoneNo: TextView

		init {
			_tvId = itemView.findViewById(android.R.id.text2)
			_tvPhoneNo = itemView.findViewById(android.R.id.text1)
		}

		fun bind(item: Phone?) {
			if (item != null) {
				_tvId.text = item.id.toString()
				_tvPhoneNo.text = item.phoneNo
			}
		}
	}

	private inner class PhoneListAdapter(diffCallback: DiffUtil.ItemCallback<Phone>) : PagingDataAdapter<Phone, PhoneViewHolder>(diffCallback) {
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhoneViewHolder {
			val inflater = LayoutInflater.from(this@MainActivity)
			val row = inflater.inflate(android.R.layout.simple_list_item_2, parent, false)
			return PhoneViewHolder(row)
		}

		override fun onBindViewHolder(holder: PhoneViewHolder, position: Int) {
			val item: Phone? = getItem(position)
			holder.bind(item)
		}
	}

	private class PhoneComparator : DiffUtil.ItemCallback<Phone>() {
		override fun areItemsTheSame(oldItem: Phone, newItem: Phone): Boolean {
			return oldItem.id == newItem.id
		}

		override fun areContentsTheSame(oldItem: Phone, newItem: Phone): Boolean {
			return oldItem == newItem
		}
	}
}
